### PR TITLE
New Events

### PR DESCRIPTION
**Pull request by JoshDreamland**
Plugs in a new YAML-based event system fundies and I have been working on.

### CODE REVIEW

File: CommandLine/emake/SOG.cpp [hunk 1]
```diff
@@ -7,54 +7,40 @@
...
-      // Needs to be big enough to accommodate the largest vk_ constant and also
-      // provide a reasonable number of alarm events, user events, etc.
-      // Using BYTE_MAX gives us both, and plenty of wiggle room.
-      for (int i = 1; i < 255; ++i) {
-        event_filenames[event_get_function_name(e.first, i) + ".edl"] = e;
-      }
-    }
-  }
+using std::map;
+using evpair = std::pair<int, int>;
```
Comment (fundies) [inline CommandLine/emake/SOG.cpp]: I would prefer a struct over std::pair<int,int> for readability but if you don't want to change itd be nice to at least comment what the ints refer to.

File: CommandLine/emake/SOG.cpp [hunk 2]
```diff
@@ -7,54 +7,40 @@
...
 
+bool ReadSOG(const std::string &input_file, Game *game) {
+  EventData event_data(ParseEventFile("events.res"));
```
Comment (fundies) [inline CommandLine/emake/SOG.cpp]: Shouldn't this be reading events.ey here?

File: CompilerSource/Makefile [hunk 3]
```diff
@@ -66,6 +67,9 @@ clean:
 $(TARGET): $(OBJECTS)
 	$(CXX) $(LDFLAGS) -o $@ $(OBJECTS) $(LDLIBS)
 
+emake:
```
Comment (fundies) [inline CompilerSource/Makefile]: we don't need a rule to build emake in the compiler as theres one in the root makefile

File: CompilerSource/compiler/components/write_defragged_events.cpp [hunk 4]
```diff
@@ -37,14 +38,49 @@ using namespace std;
...
-  void f2(int m, int i)  { id = i, mid = m; }
-  void inc(int m, int i) { mid = m, id = i, count++; }
-  void operator++(int) { count++; }
+struct EventGroup {
+  const EventDescriptor *base_event = nullptr;
+  int count = 0;
+  EventGroup &operator+=(const Event *ev) {
+    if (!base_event) {
+      base_event = ev;
+    } else if (base_event != ev) {
```
Comment (fundies) [inline CompilerSource/compiler/components/write_defragged_events.cpp]: glad you error checked that

File: CompilerSource/compiler/components/write_defragged_events.cpp [hunk 5]
```diff
@@ -37,14 +38,49 @@ using namespace std;
...
+  }
+  const EventDescriptor *operator->() const { return base_event; }
+  EventGroup(const Event *e): base_event(e) {}
+  EventGroup() = default;
+};
+
+struct UsedEventIndex {
+  map<string, EventDescriptor> base_methods;
+  set<Event> all;
+  void insert(int mid, int id) {
```
Comment (fundies) [inline CompilerSource/compiler/components/write_defragged_events.cpp]: bertter names or a note here on what "mid" is versus "id" wouldn't hurt

File: CompilerSource/compiler/components/write_defragged_events.cpp [hunk 6]
```diff
@@ -89,20 +121,28 @@ int lang_CPP::compile_writeDefraggedEvents(const GameData &game, const ParsedObj
...
+                                               obj->events[j].id);
+          if (ev2.internal_id == event.internal_id) {
+            exists = true;
+            break;
+          }
+        }
+        if (!exists) {
+          std::cout << "EVENT SYSTEM: Adding a " << event.HumanName()
+                    << " event with default code.\n";
+          auto hack = reverse_lookup_legacy_event(event);
```
Comment (fundies) [inline CompilerSource/compiler/components/write_defragged_events.cpp]: Instead of just naming the variable "hack" you should probably put a proper FIXME: here with explanation in case someone other than you needs to fo fix this

File: CompilerSource/compiler/components/write_defragged_events.cpp [hunk 7]
```diff
@@ -114,17 +154,18 @@ int lang_CPP::compile_writeDefraggedEvents(const GameData &game, const ParsedObj
...
-              if (event_has_default_code(it->second.mid,it->second.id))
-                wto << endl << "    {" << endl << "  " << event_get_default_code(it->second.mid,it->second.id) << endl << (e_is_inst ? "    }" : "    return 0;\n    }") << endl;
-              else wto << (e_is_inst ? " { } // No default " : " { return 0; } // No default ") << event_get_human_name(it->second.mid,it->second.id) << " code." << endl;
-            }
+  for (const auto &str_ev : used_events.base_methods) {
+    const string &fname = str_ev.first;
+    const EventDescriptor &event = str_ev.second;
+    const bool e_is_inst = event.IsInstance();
+    if (event.HasSubCheck() && !e_is_inst) {
+      wto << "    inline virtual bool myevent_" << fname << "_subcheck() { return false; }\n";
```
Comment (fundies) [inline CompilerSource/compiler/components/write_defragged_events.cpp]: I know this probably the best and maybe the only way to generate our code but it's so damn ugly. I wish we had some sort of template files we could subst instead so it was clearer what the end file should look like

File: CompilerSource/compiler/components/write_object_data.cpp [hunk 8]
```diff
@@ -429,13 +431,14 @@ static inline void write_object_unlink(std::ostream &wto, parsed_object *object,
...
+    if (!event_is_instance(event.mainId, event.id)) { //...well, all events which aren't stacked
+      if (event_has_iterator_declare_code(event.mainId, event.id)) {
+        if (!iscomment(event_get_iterator_declare_code(event.mainId, event.id)))
+          wto << "      " << event_get_iterator_declare_code(event.mainId, event.id) << ";\n";
       } else
-        wto << "      enigma::inst_iter *ENOBJ_ITER_myevent_" << event_get_function_name(object->events[*it].mainId, object->events[*it].id) << ";\n";
+        wto << "      enigma::inst_iter *ENOBJ_ITER_myevent_" << event_get_function_name(event.mainId, event.id) << ";\n";
     }
   }
   for (map<int, vector<int> >::iterator it = evgroup.begin(); it != evgroup.end(); it++) { // The stacked ones should have their root exported
```
Comment (fundies) [inline CompilerSource/compiler/components/write_object_data.cpp]: This smells like a robertmap you may want to mark it to make next refactor easier

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp [hunk 9]
```diff
@@ -69,6 +69,9 @@ gs_scalar viewport_x, viewport_y, viewport_w, viewport_h;
 
 namespace enigma {
 
+// Initialized here; incremented/decremented by instances that use it.
+long gui_used = 0;
```
Comment (fundies) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp]: why did you change this from bool to long?

File: events.res [hunk 10]
```diff
@@ -1,721 +0,0 @@
-# This file contains a list of all events in ENIGMA and the order in which they are executed.
```
Comment (fundies) [inline events.res]: There are 2 more copies of this file in the codebase. ones in / and is something like localeventres and the other is next to the event_reader code

File: shared/ProtoYaml/proto-yaml.h [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+#include <google/protobuf/message.h>
+
+#include <algorithm>
+#include <iostream>
+#include <fstream>
+#include <type_traits>
+#include <map>
+#include <set>
+
+namespace egm {
```
Comment (fundies) [inline shared/ProtoYaml/proto-yaml.h]: I'm not sure this is really part of the egm namespace as we reserved gmk:: & gmx:: for functions like loadproject() / writeproject(). There's no real harm in putting it under egm namespace but this is more general purpose.

File: shared/ProtoYaml/proto-yaml.cpp [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+  const proto::FieldDescriptor* field(const std::string &n) {
+    if (auto it = fields.find(n); it != fields.end()) return it->second;
+    if (auto it = fields.find(ToLower(n)); it != fields.end()) {
+      fields.insert({n, it->second});
+      return it->second;
+    }
+    return nullptr;
+  }
+
+  void Prime(const proto::Descriptor *desc) {
```
Comment (fundies) [inline shared/ProtoYaml/proto-yaml.cpp]: It's unclear to me what this function does based on it's name alone

File: shared/event_reader/event_parser.cpp [hunk 13]
```diff
@@ -1,4 +1,4 @@
-/** Copyright (C) 2008 Josh Ventura
+/** Copyright (C) 2008-2018 Josh Ventura
```
Comment (fundies) [inline shared/event_reader/event_parser.cpp]: you realize it's 2020? These copyrights everywhere are silly

File: shared/event_reader/event_parser.cpp [hunk 14]
```diff
@@ -16,612 +16,626 @@
...
-      else if (lc(str,"Case"))
-      {
-        if (last->mode != et_special and last->mode != et_spec_sys) {
-          printf("Error on line %d: Conditional attribute `case` incompatible with mode `%s`.\n",linenum,e_type_strs[last->mode]);
-          return 1;
-        } last->mmod = atol(v.c_str());
-      }
-      else {
-        printf("Warning, line %d: Ignoring unknown attribute `%s`.\n",linenum,str.c_str());
+      hacky_reverse_mapping_[cev.internal_id].main_id = main_id;
```
Comment (fundies) [inline shared/event_reader/event_parser.cpp]: TODO:

File: shared/event_reader/event_parser.cpp [hunk 15]
```diff
@@ -16,612 +16,626 @@
...
 }
 
-string event_stacked_get_root_name(int mid) {
-  main_event_info &mei = main_event_infos[mid];
-  return autoparam(mei.specs[0]->name,"stackroot");
+
+// -----------------------------------------------------------------------------
+// Legacy method implementations -----------------------------------------------
+// -----------------------------------------------------------------------------
+
```
Comment (fundies) [inline shared/event_reader/event_parser.cpp]: more todos. Perhaps best to namespace this as deprecated/hacks or somthing

File: shared/event_reader/event_parser.h [hunk 16]
```diff
@@ -25,77 +25,179 @@
...
+  // Decodes an Event ID string, such as Keyboard[Left], into an Event object.
+  Event DecodeEventString(const std::string &evstring);
+
+  EventData(buffers::config::EventFile&&);
+
+ private:
+  // Raw data read from the event configuration.
+  buffers::config::EventFile event_file_;
+  // Map of type and constant name pair (eg, {"key", "enter"}) to alias info.
+  std::map<std::pair<std::string, std::string>,
```
Comment (fundies) [inline shared/event_reader/event_parser.h]: again, im not biggest fan of std::pair but atleast you comment what the pairs are here

File: shared/event_reader/event_parser.h [hunk 17]
```diff
@@ -25,77 +25,179 @@
...
+buffers::config::EventFile ParseEventFile(std::istream &file);
 
+// Reads the given file in as an EventFile proto.
+// This is just a convenience method around ReadYamlFileAs<EventFile>().
+buffers::config::EventFile ParseEventFile(const std::string &filename);
+
+// TODO: Delete all methods below this line.
+// Let an object own this state instead of relying on statics.
+
+using std::string;
```
Comment (fundies) [inline shared/event_reader/event_parser.h]: using in headers is frowned upon

File: shared/proto_util.h [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+      if (val != NULL)
+        refl->SetEnum(msg, field, val);
+      break;
+    }
+    case CppType::CPPTYPE_STRING: {
+      refl->SetString(msg, field, value);
+      break;
+    }
+    default:
+      // error
```
Comment (fundies) [inline shared/proto_util.h]: This shouldn't happen but perhaps we should put an actual error message here just in case

File: shared/proto_util.h [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+
+inline void SetProtoField(google::protobuf::Message* msg,
+                          const google::protobuf::FieldDescriptor* field,
+                          double value) {
+  const google::protobuf::Reflection* refl = msg->GetReflection();
+  using CppType = google::protobuf::FieldDescriptor::CppType;
+
+  switch (field->cpp_type()) {
+    case CppType::CPPTYPE_INT32: {
+      refl->SetInt32(msg, field, (int) value);
```
Comment (fundies) [inline shared/proto_util.h]: some warnings will complain about c-style casts. I would prefer c++ casts here but its just a nit

File: shared/proto_util.h [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+    case CppType::CPPTYPE_INT32:  return refl->GetInt32 (msg, field);
+    case CppType::CPPTYPE_INT64:  return refl->GetInt64 (msg, field);
+    case CppType::CPPTYPE_UINT32: return refl->GetUInt32(msg, field);
+    case CppType::CPPTYPE_UINT64: return refl->GetUInt32(msg, field);
+    case CppType::CPPTYPE_DOUBLE: return refl->GetDouble(msg, field);
+    case CppType::CPPTYPE_FLOAT:  return refl->GetFloat (msg, field);
+    case CppType::CPPTYPE_BOOL:   return refl->GetBool  (msg, field);
+    case CppType::CPPTYPE_ENUM:   return refl->GetEnum(msg, field)->number();
+    case CppType::CPPTYPE_STRING: return std::stod(refl->GetString(msg, field));
+    default: return 0;
```
Comment (fundies) [inline shared/proto_util.h]: Should also probably message when this happens as it may cause weird behavior elsewhere

File: shared/strings_util.h [hunk 21]
```diff
@@ -4,6 +4,61 @@
...
+  for (char &c : snake) if (c == '_') c = ' ';
+  return snake;
+}
+
+inline std::string Capitalize(std::string str) {
+  if (char &c = str[0]; c >= 'a' && c <= 'z') c -= 'a' - 'A';
+  return str;
+}
+
+inline bool ParseBool(const std::string &b) {
```
Comment (fundies) [inline shared/strings_util.h]: should probably accept 0/1 here too

File: shared/strings_util.h [hunk 22]
```diff
@@ -4,6 +4,61 @@
...
+inline std::string StripChar(std::string str, char c) {
+  size_t i = 0;
+  for (size_t j = 0; j < str.length(); ++j) {
+    if (str[j] != c) str[i++] = str[j];
+  }
+  str.resize(i);
+  return str;
+}
 
 inline std::string string_replace_all(std::string str, std::string substr, std::string nstr) {
```
Comment (fundies) [inline shared/strings_util.h]: This looks bit outta place with its gml style naming. We should probably refactor this one's name as im pretty sure we just wrap this function again inside ::enigma_user

File: CompilerSource/Makefile (hunk 3 above)
Comment (fundies) [inline CompilerSource/Makefile]: dont need this rule

File: CommandLine/emake/Makefile [hunk 23]
```diff
@@ -1,13 +1,13 @@
-include ../../Config.mk
```
Comment (fundies) [inline CommandLine/emake/Makefile]: Messed up here

File: CommandLine/emake/Makefile [hunk 24]
```diff
@@ -1,13 +1,13 @@
...
 
 SHARED_SRC_DIR := ../../shared
 include $(SHARED_SRC_DIR)/Makefile
 include $(SHARED_SRC_DIR)/eyaml/Makefile
 include $(SHARED_SRC_DIR)/event_reader/Makefile
-include $(SHARED_SRC_DIR)/ProtoYaml/Makefile
-
-PROTO_DIR := $(SHARED_SRC_DIR)/protos/.eobjs
 
+OS := $(shell uname -s)
```
Comment (fundies) [inline CommandLine/emake/Makefile]: bad merge here too

File: CommandLine/emake/Makefile [hunk 25]
```diff
@@ -16,32 +16,59 @@ else
...
-endif
+rwildcard=$(wildcard $1/$2) $(foreach d,$(wildcard $1/*),$(call rwildcard,$d,$2))
+SOURCES := $(call rwildcard,$(SRC_DIR),*.cpp)
 
 #ON BY DEFAULT
 ifneq ($(CLI_ENABLE_EGM), FALSE)
-	CXXFLAGS += -DCLI_ENABLE_EGM
-	CXXFLAGS += -I../libEGM -I../libEGM
-	LDFLAGS += -lEGM
+	override CXXFLAGS += -DCLI_ENABLE_EGM
```
Comment (fundies) [inline CommandLine/emake/Makefile]: bad merge here

File: CommandLine/emake/Makefile [hunk 26]
```diff
@@ -16,32 +16,59 @@ else
...
+	override CXXFLAGS += -I../libEGM -I../libEGM
+	override LDFLAGS += -lEGM
 endif
 
 #OFF BY DEFAULT
 ifeq ($(CLI_ENABLE_SERVER), TRUE)
-	CXXFLAGS += -DCLI_ENABLE_SERVER
-	LDFLAGS += -lgrpc++
+	override CXXFLAGS += -DCLI_ENABLE_SERVER
```
Comment (fundies) [inline CommandLine/emake/Makefile]: bad merge

File: CommandLine/emake/Makefile [hunk 27]
```diff
@@ -16,32 +16,59 @@ else
...
+	override LDFLAGS += -lgrpc++
 else
 	SOURCES := $(filter-out ./Server.cpp, $(SOURCES))
 endif
 
-include ../../Default.mk
+OBJECTS := $(patsubst $(SRC_DIR)/%, $(OBJ_DIR)/%, $(patsubst %.cpp, %.o, $(SOURCES)))
```
Comment (fundies) [inline CommandLine/emake/Makefile]: this file seems entirely wrong. just paste the one from master

File: Makefile [hunk 28]
```diff
@@ -10,7 +10,6 @@ clean: .FORCE
 	$(MAKE) -C CompilerSource/ clean
 	$(MAKE) -C CommandLine/emake/ clean
 	$(MAKE) -C CommandLine/libEGM/ clean
-	$(MAKE) -C CommandLine/emake-tests/ clean
```
Comment (fundies) [inline Makefile]: bad merge here too

**Comment by fundies** [comment]
my vm puts -fPIC idk why it dont here. try adding ``` -DCMAKE_CXX_FLAGS="-fPIC``` to the cmake cmd

**Comment by fundies** [comment]
Can close this pr as its gone stale until 2025

File: CommandLine/emake/Makefile [hunk 29]
```diff
@@ -6,6 +6,7 @@ SHARED_SRC_DIR := ../../shared
 include $(SHARED_SRC_DIR)/Makefile
 include $(SHARED_SRC_DIR)/eyaml/Makefile
```
[the remark this answers is not in the document]
Comment (RobertBColton) [reply CommandLine/emake/Makefile]: Didn't you guys end up doing this?

File: CommandLine/emake/SOG.cpp (hunk 2 above)
Comment (RobertBColton) [reply CommandLine/emake/SOG.cpp]: What is events.ey?

File: CompilerSource/compiler/components/write_defragged_events.cpp (hunk 5 above)
Comment (RobertBColton) [reply CompilerSource/compiler/components/write_defragged_events.cpp]: Being a bit picky fundies, mid is main id. But documentation never hurts, it can only help.

File: CompilerSource/compiler/components/write_defragged_events.cpp (hunk 7 above)
Comment (RobertBColton) [reply CompilerSource/compiler/components/write_defragged_events.cpp]: That's totally doable with string interpolation. I mean even LGM does that with messages.properties.
https://stackoverflow.com/questions/37956090/string-interpolation-in-c-construct-a-stdstring-with-embedded-values-e-g

File: shared/proto_util.h (hunk 18 above)
Comment (RobertBColton) [reply shared/proto_util.h]: Need better error surfacing to the frontend, ideally we should be able to log _everything_.

File: shared/proto_util.h (hunk 19 above)
Comment (RobertBColton) [reply shared/proto_util.h]: Where did this method come from? It looks odd to me and I don't think I wrote it. Why isn't it just templated?

File: shared/strings_util.h (hunk 21 above)
Comment (RobertBColton) [reply shared/strings_util.h]: Could also add a parameter for what should be returned if the string can't be parsed.

**Comment by RobertBColton** [comment]
Why is fundies saying to close this while Josh is telling me to test it?

**Comment by RobertBColton** [comment]
Managed to reproduce this using the FPS example:
https://enigma-dev.org/edc/games.php?game=43
```
In file included from [01m[KSHELLmain.cpp:105[m[K:
[01m[KC:/Users/Owner/AppData/Local/ENIGMA/Preprocessor_Environment_Editable/IDE_EDIT_objectfunctionality.h:[m[K In member function '[01m[Kvirtual bool enigma::OBJ_obj_player::myevent_nomorehealth_subcheck()[m[K':
[01m[KC:/Users/Owner/AppData/Local/ENIGMA/Preprocessor_Environment_Editable/IDE_EDIT_objectfunctionality.h:459:10:[m[K [01;31m[Kerror: [m[K'[01m[Kout_of_health[m[K' was not declared in this scope; did you mean '[01m[K$out_of_health[m[K'?
  459 |   return [01;31m[Kout_of_health[m[K && !OoH;
      |          [01;31m[K^~~~~~~~~~~~~[m[K
      |          [32m[K$out_of_health[m[K
make[1]: *** [Makefile:177: C:/Users/Owner/AppData/Local/ENIGMA/.eobjs/Windows/Windows/Mingw_GCC_G++/Run/SHELLmain.o] Error 1
make[1]: Leaving directory '/c/Users/Owner/Documents/Projects/enigma-dev/ENIGMAsystem/SHELL'
make: *** [Makefile:22: Game] Error 2
```

**Comment by RobertBColton** [comment]
Alright 8184f9b6780cfb865d2230af4577bfd7e534919c fixed the FPS game. That was all I had found unless somebody wants me to test some other game. I tested the platform engine and a couple of my small games which were fine.

**Comment by Hughtech** [comment]
tested it on system running Manjaro 64bit.  LGM crashed and exited when trying to run an empty game. sent details and 'pastebin' texts on discord server in #troubleshooting.

**Comment by RobertBColton** [comment]
I am not able to reproduce what Hugar is reporting with an empty game over here on WIndows 10.

**Comment by time-killer-games** [comment]
Please see my complaints in the official topic and actually address the issues before merging this atrocity. kthx https://enigma-dev.org/forums/index.php?topic=3057.0

**Comment by time-killer-games** [comment]
this pr must die @JoshDreamland

**Comment by RobertBColton** [comment]
I have closed out #1479 as it seems this pull request supersedes it. I was trying to see how GMX/GMK are still working with this as it seems you haven't changed their source at all. I would appreciate clarification on that. Finally, keep an eye out for any other differences between this and my original pull request.

**Comment by time-killer-games** [comment]
I have a suggestion to add to this, I'd like us to use the exact same event name strings and id numbers used for the asynchronous extensions as yoyo does; this will give us a means to later on implement asynchronous extensions that work in both gm and enigma the same way: https://forum.yoyogames.com/index.php?threads/intro-to-multi-threading.25896/post-428547

**Comment by fundies** [comment]
@JoshDreamland  I fixed build errors in #1993. However, you seem to have broken the object in inheritance test and I'm still getting the event ordering bug in tkg's game. Also, the compiler now prints errors for our internal events like localsweep.

**Comment by fundies** [comment]
Closing in favor of #2089
